documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md

```diff
@@ -0,0 +1,366 @@
+We can add a media view at
+`/api/v1/<media>/creator/?q=<creator_name>&provider=<provider_name>`. It would
+use `self.get_queryset().filter(creator=creator, provider=provider)` to get all
+the items by the creator from the provider. This view would also need to filter
+out dead links, and return a paginated response.
+
+#### Tag page
+
+To get the items matching a tag, we can use the `tags` query parameter with a
+quoted value.
+[https://api.openverse.engineering/v1/images/?tags=%22black%20cat%22](https://api.openverse.engineering/v1/images/?source=met)
+This query returns all images with tags that contain the phrase "black cat".
+
+This would not match the tags exactly, but tags page is supposed to show all
+items that are associated with the same topic or category.
+
+We could also add a media view at `/api/v1/<media>/tags/?q=<tag_name>`, but it
+is difficult to implement because the tags are stored as a list JSONField model.
+
+### Nuxt store changes
+
+We can re-use the search store as is for these pages. Currently, the store uses
+the `searchBy` filter to determine the shape of the API query. If `searchBy`
+value is set, then the `q` parameter is replaced with the
+`<searchBy>=<searchTerm>` query parameter. API query parameters are constructed
+from the search store state in the
+[`prepare-search-query-params` method](https://github.com/WordPress/openverse/blob/b4b46b903731870015c475d2c08eebef7ec6b25b/frontend/src/utils/prepare-search-query-params.ts#L22-L25)
+
+We can update the `searchBy` filter to have one of the possible values:
+`{ searchBy: <null|creator|source|tag> }`. Then, this value would be used to
+construct the API query.
+
+#### Update the `searchBy` filter
+
+The `searchBy` filter will be used to determine the shape of the API query.
+While currently these parameters will be mutually exclusive (we can only search
+by one of them), we might want to allow searching by multiple parameters in the
+future.
+
+For other filters, we only use `toggle` method to update the value. However, for
+`searchBy`, we need to be able to check one of the `searchBy` parameters, and
+uncheck the others. To enable that, we should add a new `search` store method.
+
+This change should also update `prepare-search-query-params` to use the
+`searchBy` filter value to construct the API query:
+
+```typescript
+const prepareSearchQuery = (
+  searchParams: Record<string, string>,
+  mediaType: SupportedMediaType
+) => {
+  if (searchParams.searchBy) {
+    if (searchBy === "creator") {
+      // the name of the source is stored in the `<mediaType>Provider` filter
+      return {
+        creator: searchTerm,
+        source: searchParams[`${mediaType}Provider`],
+      }
+    } else if (searchBy === "source") {
+      return { source: searchParams[`${mediaType}Provider`] }
+    } else if (searchBy === "tag") {
+      // The parameter is plural in the API
+      return { tags: searchTerm }
+    }
+  } else {
+    // Current search query transform
+  }
+}
```

Comment:
Just noting that regardless of whether we use my fully route based approach above, we'll still need to determine the path pattern to use. Should this function also handle that, maybe returning either an array of path segments (including holes for parameters and a final object for query parameters if still needed) or the query parameter object to send to the default `search` route?